File: patch/modules/terminal/backend/filesystem.py
"""
Virtual File System - Server-side implementation
Provides a simulated filesystem for the terminal module.
"""

from typing import Literal, Optional
from .models import FileInfo


class VirtualFileSystem:
    """
    Virtual filesystem that simulates a classic computer directory structure.
    Supports both Unix-style (/) and Windows-style (\\) paths.
    """

    def __init__(self):
        # Pre-defined directory structure
        self._tree: dict[str, list[FileInfo]] = {
            # Unix-style paths (Linux/Mac)
            "/": [
                FileInfo(name="home", type="directory"),
                FileInfo(name="usr", type="directory"),
                FileInfo(name="etc", type="directory"),
                FileInfo(name="var", type="directory"),
                FileInfo(name="tmp", type="directory"),
            ],
            "/home": [
                FileInfo(name="user", type="directory"),
            ],
            "/home/user": [
                FileInfo(name="Documents", type="directory"),
                FileInfo(name="Downloads", type="directory"),
                FileInfo(name="Pictures", type="directory"),
                FileInfo(name="Music", type="directory"),
                FileInfo(name=".bashrc", type="file", size=3526),
                FileInfo(name=".profile", type="file", size=807),
                FileInfo(name="readme.txt", type="file", size=1024),
            ],
            "/home/user/Documents": [
                FileInfo(name="notes.txt", type="file", size=2048),
                FileInfo(name="project", type="directory"),
            ],
            "/home/user/Documents/project": [
                FileInfo(name="main.c", type="file", size=4096),
                FileInfo(name="Makefile", type="file", size=512),
            ],
            "/home/user/Downloads": [
                FileInfo(name="setup.exe", type="file", size=1048576),
            ],
            "/home/user/Pictures": [
                FileInfo(name="photo.jpg", type="file", size=524288),
            ],
            "/home/user/Music": [
                FileInfo(name="song.mp3", type="file", size=3145728),
            ],
            "/usr": [
                FileInfo(name="bin", type="directory"),
                FileInfo(name="lib", type="directory"),
                FileInfo(name="share", type="directory"),
            ],
            "/usr/bin": [
                FileInfo(name="ls", type="file", size=133792),
                FileInfo(name="cd", type="file", size=0),
            ],
            "/etc": [
                FileInfo(name="passwd", type="file", size=2048),
                FileInfo(name="hosts", type="file", size=256),
            ],
            "/var": [
                FileInfo(name="log", type="directory"),
            ],
            "/var/log": [
                FileInfo(name="syslog", type="file", size=102400),
            ],
            "/tmp": [],
            # macOS specific
            "/Users": [
                FileInfo(name="user", type="directory"),
            ],
            "/Users/user": [
                FileInfo(name="Documents", type="directory"),
                FileInfo(name="Downloads", type="directory"),
                FileInfo(name="Desktop", type="directory"),
                FileInfo(name="Pictures", type="directory"),
                FileInfo(name="Music", type="directory"),
                FileInfo(name=".zshrc", type="file", size=1024),
            ],
            "/Users/user/Documents": [
                FileInfo(name="notes.txt", type="file", size=2048),
            ],
            "/Users/user/Downloads": [],
            "/Users/user/Desktop": [
                FileInfo(name="readme.txt", type="file", size=512),
            ],
            "/Users/user/Pictures": [],
            "/Users/user/Music": [],
            # Windows-style paths (DOS/Win95)
            "C:\\": [
                FileInfo(name="WINDOWS", type="directory"),
                FileInfo(name="PROGRA~1", type="directory"),
                FileInfo(name="DOS", type="directory"),
                FileInfo(name="AUTOEXEC.BAT", type="file", size=256),
                FileInfo(name="CONFIG.SYS", type="file", size=128),
                FileInfo(name="COMMAND.COM", type="file", size=54619),
            ],
            "C:\\WINDOWS": [
                FileInfo(name="SYSTEM", type="directory"),
                FileInfo(name="SYSTEM32", type="directory"),
                FileInfo(name="TEMP", type="directory"),
                FileInfo(name="WIN.INI", type="file", size=4096),
                FileInfo(name="SYSTEM.INI", type="file", size=2048),
            ],
            "C:\\WINDOWS\\SYSTEM": [
                FileInfo(name="USER.EXE", type="file", size=264016),
                FileInfo(name="KERNEL.EXE", type="file", size=327680),
            ],
            "C:\\WINDOWS\\SYSTEM32": [
                FileInfo(name="DRIVERS", type="directory"),
            ],
            "C:\\WINDOWS\\SYSTEM32\\DRIVERS": [],
            "C:\\WINDOWS\\TEMP": [],
            "C:\\PROGRA~1": [
                FileInfo(name="ACCESSORIES", type="directory"),
            ],
            "C:\\PROGRA~1\\ACCESSORIES": [
                FileInfo(name="NOTEPAD.EXE", type="file", size=32768),
                FileInfo(name="CALC.EXE", type="file", size=65536),
            ],
            "C:\\DOS": [
                FileInfo(name="EDIT.COM", type="file", size=413),
                FileInfo(name="QBASIC.EXE", type="file", size=194309),
                FileInfo(name="HELP.COM", type="file", size=413),
            ],
        }

    def _normalize_path(self, path: str, separator: str) -> str:
        """Normalize a path for consistent lookup."""
        if not path or path == separator:
            return "C:\\" if separator == "\\" else "/"

        normalized = path.rstrip(separator) if len(path) > 1 else path

        if separator == "\\":
            normalized = normalized.upper()

        return normalized

    def list_directory(self, path: str) -> Optional[list[FileInfo]]:
        """List contents of a directory."""
        # Try exact match first
        if path in self._tree:
            return self._tree[path]

        # Try uppercase (for Windows paths)
        upper_path = path.upper()
        if upper_path in self._tree:
            return self._tree[upper_path]

        # Try without trailing separator
        stripped = path.rstrip("/\\")
        if stripped in self._tree:
            return self._tree[stripped]

        return None

    def exists(self, path: str) -> bool:
        """Check if a path exists."""
        if self.list_directory(path) is not None:
            return True

        # Check if it's a file in parent directory
        separator = "\\" if "\\" in path else "/"
        parts = path.split(separator)
        filename = parts[-1]
        parent_path = separator.join(parts[:-1]) or (
            "C:\\" if separator == "\\" else "/"
        )

        parent_contents = self.list_directory(parent_path)
        if parent_contents and filename:
            return any(
                f.name == filename or f.name.upper() == filename.upper()
                for f in parent_contents
            )

        return False

    def is_directory(self, path: str) -> bool:
        """Check if a path is a directory."""
        return self.list_directory(path) is not None

    def resolve_path(
        self,
        cwd: str,
        target: str,
        separator: Literal["/", "\\"]
    ) -> str:
        """Resolve a relative path to an absolute path."""
        if not target or target == ".":
            return cwd

        # Check if already absolute
        is_absolute = (
            target.startswith("/") if separator == "/"
            else bool(target and len(target) >= 2 and target[1] == ":")
        )

        if is_absolute:
            return self._normalize_path(target, separator)

        # Handle relative path
        parts = cwd.split(separator)
        # For Windows, preserve drive letter
        if separator == "\\" and parts:
            base_parts = [parts[0]]
            parts = parts[1:]
        else:
            base_parts = []

        result_parts = [p for p in parts if p]
        target_parts = [p for p in target.split(separator) if p]

        for part in target_parts:
            if part == "..":
                if result_parts:
                    result_parts.pop()
            elif part != ".":
                result_parts.append(part)

        if separator == "\\":
            drive = base_parts[0] if base_parts else "C:"
            return f"{drive}\\{separator.join(result_parts)}" if result_parts else f"{drive}\\"

        return "/" + "/".join(result_parts)


# Global instance
virtual_fs = VirtualFileSystem()

File: patch/modules/terminal/frontend/filesystem.ts
/**
 * Virtual File System - Client-side implementation
 * Simulates a classic file system for the terminal
 */

import type { VirtualFile, VirtualFileSystem } from './types';

// Pre-defined directory structure
const directoryTree: Record<string, VirtualFile[]> = {
  // Unix-style paths (Linux/Mac)
  '/': [
    { name: 'home', type: 'directory' },
    { name: 'usr', type: 'directory' },
    { name: 'etc', type: 'directory' },
    { name: 'var', type: 'directory' },
    { name: 'tmp', type: 'directory' },
  ],
  '/home': [
    { name: 'user', type: 'directory' },
  ],
  '/home/user': [
    { name: 'Documents', type: 'directory' },
    { name: 'Downloads', type: 'directory' },
    { name: 'Pictures', type: 'directory' },
    { name: 'Music', type: 'directory' },
    { name: '.bashrc', type: 'file', size: 3526 },
    { name: '.profile', type: 'file', size: 807 },
    { name: 'readme.txt', type: 'file', size: 1024 },
  ],
  '/home/user/Documents': [
    { name: 'notes.txt', type: 'file', size: 2048 },
    { name: 'project', type: 'directory' },
  ],
  '/home/user/Documents/project': [
    { name: 'main.c', type: 'file', size: 4096 },
    { name: 'Makefile', type: 'file', size: 512 },
  ],
  '/home/user/Downloads': [
    { name: 'setup.exe', type: 'file', size: 1048576 },
  ],
  '/home/user/Pictures': [
    { name: 'photo.jpg', type: 'file', size: 524288 },
  ],
  '/home/user/Music': [
    { name: 'song.mp3', type: 'file', size: 3145728 },
  ],
  '/usr': [
    { name: 'bin', type: 'directory' },
    { name: 'lib', type: 'directory' },
    { name: 'share', type: 'directory' },
  ],
  '/usr/bin': [
    { name: 'ls', type: 'file', size: 133792 },
    { name: 'cd', type: 'file', size: 0 },
  ],
  '/etc': [
    { name: 'passwd', type: 'file', size: 2048 },
    { name: 'hosts', type: 'file', size: 256 },
  ],
  '/var': [
    { name: 'log', type: 'directory' },
  ],
  '/var/log': [
    { name: 'syslog', type: 'file', size: 102400 },
  ],
  '/tmp': [],

  // macOS specific
  '/Users': [
    { name: 'user', type: 'directory' },
  ],
  '/Users/user': [
    { name: 'Documents', type: 'directory' },
    { name: 'Downloads', type: 'directory' },
    { name: 'Desktop', type: 'directory' },
    { name: 'Pictures', type: 'directory' },
    { name: 'Music', type: 'directory' },
    { name: '.zshrc', type: 'file', size: 1024 },
  ],
  '/Users/user/Documents': [
    { name: 'notes.txt', type: 'file', size: 2048 },
  ],
  '/Users/user/Downloads': [],
  '/Users/user/Desktop': [
    { name: 'readme.txt', type: 'file', size: 512 },
  ],
  '/Users/user/Pictures': [],
  '/Users/user/Music': [],

  // Windows-style paths (DOS/Win95)
  'C:\\': [
    { name: 'WINDOWS', type: 'directory' },
    { name: 'PROGRA~1', type: 'directory' },
    { name: 'DOS', type: 'directory' },
    { name: 'AUTOEXEC.BAT', type: 'file', size: 256 },
    { name: 'CONFIG.SYS', type: 'file', size: 128 },
    { name: 'COMMAND.COM', type: 'file', size: 54619 },
  ],
  'C:\\WINDOWS': [
    { name: 'SYSTEM', type: 'directory' },
    { name: 'SYSTEM32', type: 'directory' },
    { name: 'TEMP', type: 'directory' },
    { name: 'WIN.INI', type: 'file', size: 4096 },
    { name: 'SYSTEM.INI', type: 'file', size: 2048 },
  ],
  'C:\\WINDOWS\\SYSTEM': [
    { name: 'USER.EXE', type: 'file', size: 264016 },
    { name: 'KERNEL.EXE', type: 'file', size: 327680 },
  ],
  'C:\\WINDOWS\\SYSTEM32': [
    { name: 'DRIVERS', type: 'directory' },
  ],
  'C:\\WINDOWS\\SYSTEM32\\DRIVERS': [],
  'C:\\WINDOWS\\TEMP': [],
  'C:\\PROGRA~1': [
    { name: 'ACCESSORIES', type: 'directory' },
  ],
  'C:\\PROGRA~1\\ACCESSORIES': [
    { name: 'NOTEPAD.EXE', type: 'file', size: 32768 },
    { name: 'CALC.EXE', type: 'file', size: 65536 },
  ],
  'C:\\DOS': [
    { name: 'EDIT.COM', type: 'file', size: 413 },
    { name: 'QBASIC.EXE', type: 'file', size: 194309 },
    { name: 'HELP.COM', type: 'file', size: 413 },
  ],
};

export function createVirtualFileSystem(): VirtualFileSystem {
  const normalizePath = (path: string, separator: '/' | '\\'): string => {
    // Handle empty or just separator
    if (!path || path === separator) {
      return separator === '\\' ? 'C:\\' : '/';
    }

    // Remove trailing separator (except for root)
    let normalized = path;
    if (normalized.length > 1 && normalized.endsWith(separator)) {
      normalized = normalized.slice(0, -1);
    }

    // Ensure Windows paths are uppercase for consistency
    if (separator === '\\') {
      normalized = normalized.toUpperCase();
    }

    return normalized;
  };

  const getAbsolutePath = (cwd: string, target: string, separator: '/' | '\\'): string => {
    if (!target || target === '.') {
      return cwd;
    }

    const isAbsolute = separator === '/'
      ? target.startsWith('/')
      : /^[A-Za-z]:\\/.test(target);

    if (isAbsolute) {
      return normalizePath(target, separator);
    }

    // Handle .. and relative paths
    const parts = cwd.split(separator).filter(Boolean);
    const targetParts = target.split(separator).filter(Boolean);

    // For Windows, keep drive letter
    let baseParts: string[] = [];
    if (separator === '\\' && parts.length > 0) {
      baseParts = [parts[0]]; // Keep C:
      parts.shift();
    }

    const resultParts = [...parts];

    for (const part of targetParts) {
      if (part === '..') {
        if (resultParts.length > 0) {
          resultParts.pop();
        }
      } else if (part !== '.') {
        resultParts.push(part);
      }
    }

    if (separator === '\\') {
      const drive = baseParts[0] || 'C:';
      return resultParts.length === 0
        ? `${drive}\\`
        : `${drive}\\${resultParts.join('\\')}`;
    }

    return '/' + resultParts.join('/');
  };

  const resolvePath = (cwd: string, target: string, separator: '/' | '\\'): string => {
    return getAbsolutePath(cwd, target, separator);
  };

  const listDirectory = (path: string): VirtualFile[] | null => {
    const normalized = path.toUpperCase().replace(/\/$/, '') || '/';

    // Try exact match first
    if (directoryTree[path]) {
      return directoryTree[path];
    }

    // Try uppercase (for Windows paths)
    if (directoryTree[normalized]) {
      return directoryTree[normalized];
    }

    // Try with trailing separator removed
    const withoutTrailing = path.replace(/[/\\]$/, '');
    if (directoryTree[withoutTrailing]) {
      return directoryTree[withoutTrailing];
    }

    return null;
  };

  const exists = (path: string): boolean => {
    // Check if it's a directory
    if (listDirectory(path) !== null) {
      return true;
    }

    // Check if it's a file in parent directory
    const separator = path.includes('\\') ? '\\' : '/';
    const parts = path.split(separator);
    const fileName = parts.pop();
    const parentPath = parts.join(separator) || (separator === '/' ? '/' : 'C:\\');

    const parentContents = listDirectory(parentPath);
    if (parentContents && fileName) {
      return parentContents.some(f => f.name === fileName || f.name.toUpperCase() === fileName.toUpperCase());
    }

    return false;
  };

  const isDirectory = (path: string): boolean => {
    return listDirectory(path) !== null;
  };

  return {
    listDirectory,
    exists,
    isDirectory,
    resolvePath,
    getAbsolutePath,
  };
}

export const virtualFS = createVirtualFileSystem();
